File: adaptors/mdast/mdast.py
from __future__ import annotations

from typing import Any, Dict, List

from ast_parser import Document, Heading, Paragraph, ParagraphLine, Comment, MarkdownBlock, Node


def document_to_mdast(document: Document) -> Dict[str, Any]:
    """Convert a Spaceup Document into an mdast-compatible dict tree."""

    def clamp_heading_depth(level: int) -> int:
        return 6 if level > 6 else 1 if level < 1 else level

    def text_node(value: str) -> Dict[str, Any]:
        return {"type": "text", "value": value}

    def html_comment_node(value: str) -> Dict[str, Any]:
        return {"type": "html", "value": f"<!-- {value} -->"}

    def break_node() -> Dict[str, Any]:
        return {"type": "break"}

    def inline_text_to_children(text: str) -> List[Dict[str, Any]]:
        # Minimal mapping: treat entire inline Markdown as plain text.
        # Future: parse markdown-it tokens and map to mdast phrasing nodes.
        if not text:
            return []
        return [text_node(text)]

    def paragraph_lines_are_bulleted(lines: List[ParagraphLine]) -> bool:
        if not lines:
            return False
        for line in lines:
            if not line.content.text.lstrip().startswith("- "):
                return False
        return True

    def list_from_bulleted_paragraph(paragraph: Paragraph) -> Dict[str, Any]:
        items: List[Dict[str, Any]] = []
        for line in paragraph.lines:
            # Strip a single leading bullet "- "
            text = line.content.text.lstrip()[2:].strip()
            item_children: List[Dict[str, Any]] = []
            para_children = inline_text_to_children(text)
            if line.inline_comment:
                para_children.append(html_comment_node(line.inline_comment))
            item_children.append({"type": "paragraph", "children": para_children})
            items.append({"type": "listItem", "children": item_children})
        return {"type": "list", "ordered": False, "spread": False, "children": items}

    def paragraph_to_mdast(paragraph: Paragraph) -> Dict[str, Any]:
        # Special-case: simple unordered list using "- " prefix on all lines
        if paragraph_lines_are_bulleted(paragraph.lines):
            return list_from_bulleted_paragraph(paragraph)

        children: List[Dict[str, Any]] = []
        for idx, line in enumerate(paragraph.lines):
            children.extend(inline_text_to_children(line.content.text))
            if line.inline_comment:
                children.append(html_comment_node(line.inline_comment))
            # Insert a soft break between lines (not after the last)
            if idx < len(paragraph.lines) - 1:
                children.append(break_node())
        return {"type": "paragraph", "children": children}

    def heading_to_mdast(heading: Heading) -> Dict[str, Any]:
        children = inline_text_to_children(heading.content.text)
        node: Dict[str, Any] = {
            "type": "heading",
            "depth": clamp_heading_depth(heading.level),
            "children": children,
        }
        if heading.level > 6:
            node["data"] = {"spaceupHeadingLevel": heading.level}
        return node

    def node_to_mdast(node: Node) -> Dict[str, Any] | None:
        if isinstance(node, Heading):
            return heading_to_mdast(node)
        if isinstance(node, Paragraph):
            return paragraph_to_mdast(node)
        if isinstance(node, Comment):
            if not node.text:
                return None
            return html_comment_node(node.text)
        if isinstance(node, MarkdownBlock):
            # Treat as a plain paragraph for now.
            return {"type": "paragraph", "children": inline_text_to_children(node.text)}
        return None

    children: List[Dict[str, Any]] = []
    for child in document.children:
        converted = node_to_mdast(child)
        if converted is None:
            continue
        children.append(converted)

    return {"type": "root", "children": children}

File: js/adaptors/mdast/mdast.js
"use strict";

function clampHeadingDepth(level) {
  if (level < 1) return 1;
  if (level > 6) return 6;
  return level;
}

function textNode(value) {
  return { type: "text", value: String(value ?? "") };
}

function htmlCommentNode(value) {
  return { type: "html", value: `<!-- ${String(value ?? "")} -->` };
}

function breakNode() {
  return { type: "break" };
}

function inlineTextToChildren(text) {
  if (!text) return [];
  return [textNode(text)];
}

function paragraphLinesAreBulleted(lines) {
  if (!Array.isArray(lines) || lines.length === 0) return false;
  for (const line of lines) {
    const t = (line?.content?.text ?? "").replace(/^\s+/, "");
    if (!t.startsWith("- ")) return false;
  }
  return true;
}

function listFromBulletedParagraph(paragraph) {
  const items = [];
  for (const line of paragraph.lines) {
    const raw = String(line?.content?.text ?? "");
    const itemText = raw.replace(/^\s+/, "").slice(2).trim();
    const paraChildren = inlineTextToChildren(itemText);
    if (line.inline_comment) {
      paraChildren.push(htmlCommentNode(line.inline_comment));
    }
    items.push({
      type: "listItem",
      spread: false,
      children: [{ type: "paragraph", children: paraChildren }],
    });
  }
  return { type: "list", ordered: false, spread: false, children: items };
}

function paragraphToMdast(paragraph) {
  if (paragraphLinesAreBulleted(paragraph.lines)) {
    return listFromBulletedParagraph(paragraph);
  }
  const children = [];
  for (let i = 0; i < paragraph.lines.length; i += 1) {
    const line = paragraph.lines[i];
    children.push(...inlineTextToChildren(line?.content?.text ?? ""));
    if (line.inline_comment) children.push(htmlCommentNode(line.inline_comment));
    if (i < paragraph.lines.length - 1) children.push(breakNode());
  }
  return { type: "paragraph", children };
}

function headingToMdast(heading) {
  const node = {
    type: "heading",
    depth: clampHeadingDepth(Number(heading?.level ?? 1)),
    children: inlineTextToChildren(heading?.content?.text ?? ""),
  };
  if (heading.level > 6) {
    node.data = { spaceupHeadingLevel: heading.level };
  }
  return node;
}

function documentToMdast(document) {
  const children = [];
  for (const node of document.children || []) {
    if (node && typeof node.level === "number" && node.content) {
      children.push(headingToMdast(node));
      continue;
    }
    if (node && Array.isArray(node.lines)) {
      children.push(paragraphToMdast(node));
      continue;
    }
    if (node && typeof node.text === "string") {
      if (node.text) children.push(htmlCommentNode(node.text));
      continue;
    }
    if (node && typeof node.text === "string" && node.tokens) {
      // MarkdownBlock (not used currently) → paragraph fallback
      children.push({ type: "paragraph", children: inlineTextToChildren(node.text) });
    }
  }
  return { type: "root", children };
}

module.exports = {
  documentToMdast,
};


